src/EditorFeatures/Core/ValueTracking/ValueTrackingService.Visitor.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ValueTracking
+{
+    internal partial class ValueTrackingService
+    {
+        private class OperationCollector
+        {
+            public ValueTrackingProgressCollector ProgressCollector { get; }
+            public Solution Solution { get; }
+
+            public OperationCollector(ValueTrackingProgressCollector progressCollector, Solution solution)
+            {
+                ProgressCollector = progressCollector;
+                Solution = solution;
+            }
+
+            public Task VisitAsync(IOperation operation, CancellationToken cancellationToken)
+                => operation switch
+                {
+                    IObjectCreationOperation objectCreationOperation => VisitObjectCreationAsync(objectCreationOperation, cancellationToken),
+                    IInvocationOperation invocationOperation => VisitInvocationAsync(invocationOperation, cancellationToken),
+                    ILiteralOperation literalOperation => VisitLiteralAsync(literalOperation, cancellationToken),
+                    IReturnOperation returnOperation => VisitReturnAsync(returnOperation, cancellationToken),
+                    IArgumentOperation argumentOperation => ShouldTrackArgument(argumentOperation) ? VisitAsync(argumentOperation.Value, cancellationToken) : Task.CompletedTask,
+                    ILocalReferenceOperation or
+                        IParameterReferenceOperation or
+                        IFieldReferenceOperation or
+                        IPropertyReferenceOperation => VisitReferenceAsync(operation, cancellationToken),
+
+                    IAssignmentOperation assignmentOperation => VisitAssignmentOperationAsync(assignmentOperation, cancellationToken),
+
+                    // All the operations in this group don't have meaningful use themselves, but the child operations
+                    // could be meaningful for value tracking. 
+                    // Ex: Binary operation of [| x + y |]
+                    // both x and y should be evaluated separately
+                    IBinaryOperation or
+                        IAwaitOperation or
+                        IBlockOperation => VisitChildrenAsync(operation, cancellationToken),
```

Comment:
this seem sa little error prone that you've mechnically enumerated these cases.  waht about things like unary ops?  what about ternary ops?  etc.   should we conservatively just always walk teh children, except for cases we kno we don't need to?

Reply:
I'm not sure one list is easier to maintain than the other, and I find this easier to follow. I'm not strongly opinionated on this though, and can move `VisitChildrenAsync` to be part of `VisitDefaultAsync`